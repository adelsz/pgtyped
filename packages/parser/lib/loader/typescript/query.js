import { CharStreams, CommonTokenStream } from 'antlr4ts';
import { ParseTreeWalker } from 'antlr4ts/tree/ParseTreeWalker.js';
import { QueryLexer } from './parser/QueryLexer.js';
import { QueryParser, } from './parser/QueryParser.js';
import { Logger } from '../sql/logger.js';
import { Interval } from 'antlr4ts/misc/index.js';
export var ParamType;
(function (ParamType) {
    ParamType["Scalar"] = "scalar";
    ParamType["Object"] = "object";
    ParamType["ScalarArray"] = "scalar_array";
    ParamType["ObjectArray"] = "object_array";
})(ParamType || (ParamType = {}));
export function assert(condition) {
    if (!condition) {
        throw new Error('Assertion Failed');
    }
}
class ParseListener {
    constructor(queryName, logger) {
        this.query = {};
        this.currentParam = {};
        this.currentSelection = {};
        this.query.name = queryName;
        this.logger = logger;
    }
    enterQuery(ctx) {
        const { inputStream } = ctx.start;
        const end = ctx.stop.stopIndex;
        const interval = new Interval(0, end);
        const text = inputStream.getText(interval);
        this.query = {
            name: this.query.name,
            text,
            params: [],
        };
    }
    enterParamName(ctx) {
        this.currentParam = {
            name: ctx.text,
            selection: undefined,
        };
    }
    enterScalarParamName(ctx) {
        const required = !!ctx.REQUIRED_MARK();
        const name = ctx.ID().text;
        this.currentParam = {
            name,
            required,
        };
    }
    exitParam(ctx) {
        const defLoc = {
            a: ctx.start.startIndex,
            b: ctx.stop.stopIndex,
            line: ctx.start.line,
            col: ctx.start.charPositionInLine,
        };
        this.currentParam.location = defLoc;
        this.currentParam.selection = this.currentSelection;
        this.query.params.push(this.currentParam);
        this.currentSelection = {};
        this.currentParam = {};
    }
    enterScalarParam() {
        this.currentSelection = {
            type: ParamType.Scalar,
        };
    }
    enterPickParam() {
        this.currentSelection = {
            type: ParamType.Object,
            keys: [],
        };
    }
    enterArrayPickParam() {
        this.currentSelection = {
            type: ParamType.ObjectArray,
            keys: [],
        };
    }
    enterArrayParam() {
        this.currentSelection = {
            type: ParamType.ScalarArray,
        };
    }
    enterPickKey(ctx) {
        assert('keys' in this.currentSelection);
        const required = !!ctx.REQUIRED_MARK();
        const name = ctx.ID().text;
        this.currentSelection.keys.push({ name, required });
    }
}
function parseText(text, queryName = 'query') {
    const logger = new Logger();
    const inputStream = CharStreams.fromString(text);
    const lexer = new QueryLexer(inputStream);
    lexer.removeErrorListeners();
    lexer.addErrorListener(logger);
    const tokenStream = new CommonTokenStream(lexer);
    const parser = new QueryParser(tokenStream);
    parser.removeErrorListeners();
    parser.addErrorListener(logger);
    const tree = parser.input();
    const listener = new ParseListener(queryName, logger);
    ParseTreeWalker.DEFAULT.walk(listener, tree);
    return {
        query: listener.query,
        events: logger.parseEvents,
    };
}
export default parseText;
//# sourceMappingURL=query.js.map