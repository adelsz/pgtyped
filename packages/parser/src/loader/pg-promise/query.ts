import { QueryParserListener } from './parser/QueryParserListener.js';
import { CharStreams, CommonTokenStream } from 'antlr4ts';
import { ParseTreeWalker } from 'antlr4ts/tree/ParseTreeWalker.js';
import { QueryLexer } from './parser/QueryLexer.js';
import {
  FormatterContext,
  ParamContext,
  ParamIndexedContext,
  ParamNamedContext,
  QueryContext,
  QueryParser,
} from './parser/QueryParser.js';
import { Logger, ParseEvent } from '../sql/logger.js';
import { Interval } from 'antlr4ts/misc/index.js';

export enum ParamType {
  Scalar = 'scalar',
  ScalarArray = 'scalar_array',
  Identifier = 'identifier', // not supported, but we give a proper error message
  Raw = 'raw', // not supported, but we give a proper error message
}

export type ParamSelection =
  | {
      type: ParamType.Scalar;
    }
  | {
      type: ParamType.ScalarArray;
    }
  | {
      type: ParamType.Identifier;
    }
  | {
      type: ParamType.Raw;
    };

export interface Param {
  name: string;
  selection: ParamSelection;
  required: boolean;
  nullable: boolean;
  location: CodeInterval;
}

interface CodeInterval {
  a: number;
  b: number;
  line: number;
  col: number;
}

export interface Query {
  name: string;
  params: Param[];
  text: string;
}

class ParserListener implements QueryParserListener {
  logger: Logger;
  query: Partial<Query> = {};
  private currentParam: Partial<Param> = {};

  constructor(queryName: string, logger: Logger) {
    this.query.name = queryName;
    this.logger = logger;
  }

  enterQuery(ctx: QueryContext) {
    const { inputStream } = ctx.start;
    const end = ctx.stop!.stopIndex;

    const interval = new Interval(0, end);
    const text = inputStream!.getText(interval);
    this.query = {
      name: this.query.name,
      text,
      params: [],
    };
  }

  enterParamNamed(ctx: ParamNamedContext) {
    this.currentParam = {
      name: ctx.ID().text,
      nullable: ctx.NULLABILITY_MARK() !== undefined,
      required: true,
      selection: { type: ParamType.Scalar },
    };
  }

  enterParamIndexed(ctx: ParamIndexedContext) {
    this.currentParam = {
      name: ctx.INTEGER().text,
      nullable: ctx.NULLABILITY_MARK() !== undefined,
      required: true,
      selection: { type: ParamType.Scalar },
    };
  }

  enterFormatter(ctx: FormatterContext) {
    const fmtText = ctx.ID()?.text;
    const fmtShort = ctx.FORMATTER_SHORT()?.text;
    let type: ParamType | undefined;
    switch (fmtText) {
      case 'list':
      case 'csv':
        type = ParamType.ScalarArray;
        break;
      case 'alias':
      case 'name':
        type = ParamType.Identifier;
        break;
      case 'raw':
        type = ParamType.Raw;
        break;
    }
    if (!type) {
      switch (fmtShort) {
        case '~':
          type = ParamType.Identifier;
          break;
        case '^':
          type = ParamType.Raw;
          break;
        default:
          type = ParamType.Scalar;
      }
    }
    this.currentParam.selection = {
      type,
    };
  }

  exitParam(ctx: ParamContext) {
    const defLoc = {
      a: ctx.start.startIndex,
      b: ctx.stop!.stopIndex,
      line: ctx.start.line,
      col: ctx.start.charPositionInLine,
    };
    this.currentParam.location = defLoc;
    this.query.params!.push(this.currentParam as Param);
    this.currentParam = {};
  }
}

export function parseTextPgPromise(
  text: string,
  queryName: string = 'query',
): { query: Query; events: ParseEvent[] } {
  const logger = new Logger();
  const inputStream = CharStreams.fromString(text);
  const lexer = new QueryLexer(inputStream);
  lexer.removeErrorListeners();
  lexer.addErrorListener(logger);
  const tokenStream = new CommonTokenStream(lexer);
  const parser = new QueryParser(tokenStream);
  parser.removeErrorListeners();
  parser.addErrorListener(logger);

  const tree = parser.input();

  const listener = new ParserListener(queryName, logger);
  ParseTreeWalker.DEFAULT.walk(listener as QueryParserListener, tree);

  return {
    query: listener.query as Query,
    events: logger.parseEvents,
  };
}
