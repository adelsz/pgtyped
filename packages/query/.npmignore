src/**/*
lib/**/*.test.*
tsconfig.json
*.tgz