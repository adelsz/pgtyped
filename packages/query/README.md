## @pgtyped/query

This package provides the `sql` tagged template.  
The `sql` tagged template requires a generic parameter: `<TQueryType>`.
For each query PgTyped generates an interface that can be used in this parameter to type your query.

To run a query defined with the `sql` tagged template, call the `sql.run` method.  
The `sql.run` method automatically enforces correct input `TParams` and output `TResult` types.

```js
public run: (
  params: TParams,
  dbConnection: IDatabaseConnection,
) => Promise<TResult[]>;
```

Here `dbConnection` is any object that satisifies the `IDatabaseConnection` interface. It is used to actually send the query to the DB for execution.

```
interface IDatabaseConnection {
  query: (query: string, bindings: any[]) => Promise<{ rows: any[] }>;
}
```

This is usually the `client` object created with [node-postgres](https://github.com/brianc/node-postgres), but can be any other connection of your choice.

This package is part of the pgtyped project.  
Refer to [README](https://github.com/adelsz/pgtyped) for details.
