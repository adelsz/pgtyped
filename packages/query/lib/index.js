export { getTypes, startup } from './actions.js';
export { isAlias, isEnum, isEnumArray, isImport, } from './type.js';
//# sourceMappingURL=index.js.map