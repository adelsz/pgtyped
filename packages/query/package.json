{
  "name": "@pgtyped/query",
  "version": "1.0.1",
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/adelsz/pgtyped.git"
  },
  "homepage": "https://github.com/adelsz/pgtyped",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "test": "jest",
    "build": "tsc --declaration",
    "check": "tsc --noEmit",
    "watch": "tsc --declaration --watch --preserveWatchOutput",
    "parsegen-sql": "antlr4ts -visitor -Xexact-output-dir -o src/loader/sql/parser src/loader/sql/grammar/*.g4",
    "parsegen-ts": "antlr4ts -visitor -Xexact-output-dir -o src/loader/typescript/parser src/loader/typescript/grammar/*.g4"
  },
  "jest": {
    "roots": [
      "src"
    ],
    "testRegex": "\\.test\\.tsx?$",
    "transform": {
      "^.+\\.tsx?$": "ts-jest"
    }
  },
  "dependencies": {
    "@pgtyped/wire": "^1.0.1",
    "@types/chalk": "^2.2.0",
    "@types/debug": "^4.1.4",
    "antlr4ts": "0.5.0-alpha.4",
    "chalk": "^4.1.0",
    "debug": "^4.1.1"
  },
  "devDependencies": {
    "antlr4ts-cli": "0.5.0-alpha.4"
  },
  "peerDependencies": {
    "typescript": "3.1 - 4"
  }
}
