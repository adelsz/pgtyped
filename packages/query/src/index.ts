export { getTypes, startup, IParseError } from './actions';

export {
  ParamTransform,
  IQueryParameters,
  IInterpolatedQuery,
} from './preprocessor';

export { processTSQueryAST } from './preprocessor-ts';
export { processSQLQueryIR } from './preprocessor-sql';

export { AsyncQueue } from '@pgtyped/wire';

export {
  default as parseTypeScriptFile,
  TSQueryAST,
} from './loader/typescript';

export {
  default as parseSQLFile,
  SQLQueryAST,
  SQLQueryIR,
  prettyPrintEvents,
  queryASTToIR,
} from './loader/sql';

export { default as sql, TaggedQuery, PreparedQuery } from './tag';
