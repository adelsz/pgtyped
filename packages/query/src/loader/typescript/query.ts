import { QueryParserListener } from './parser/QueryParserListener';
import { CharStreams, CommonTokenStream } from 'antlr4ts';
import { ParseTreeWalker } from 'antlr4ts/tree/ParseTreeWalker';
import { QueryLexer } from './parser/QueryLexer';
import {
  ArrayParamContext,
  ParamContext,
  ParamNameContext,
  PickKeyContext,
  QueryContext,
  QueryParser,
  ScalarParamNameContext,
} from './parser/QueryParser';
import { Logger, ParseEvent } from '../sql/logger';
import { Interval } from 'antlr4ts/misc';

export enum ParamType {
  Scalar = 'scalar',
  Object = 'object',
  ScalarArray = 'scalar_array',
  ObjectArray = 'object_array',
}

export interface ParamKey {
  name: string;
  required: boolean;
}

export type ParamSelection =
  | {
      type: ParamType.Scalar;
    }
  | {
      type: ParamType.ScalarArray;
    }
  | {
      type: ParamType.Object | ParamType.ObjectArray;
      keys: ParamKey[];
    };

export interface Param {
  name: string;
  selection: ParamSelection;
  required: boolean;
  location: CodeInterval;
}

interface CodeInterval {
  a: number;
  b: number;
  line: number;
  col: number;
}

export interface Query {
  name: string;
  params: Param[];
  text: string;
}

export function assert(condition: any): asserts condition {
  if (!condition) {
    throw new Error('Assertion Failed');
  }
}

class ParseListener implements QueryParserListener {
  logger: Logger;
  query: Partial<Query> = {};
  private currentParam: Partial<Param> = {};
  private currentSelection: Partial<ParamSelection> = {};

  constructor(queryName: string, logger: Logger) {
    this.query.name = queryName;
    this.logger = logger;
  }

  enterQuery(ctx: QueryContext) {
    const { inputStream } = ctx.start;
    const end = ctx.stop!.stopIndex;

    const interval = new Interval(0, end);
    const text = inputStream!.getText(interval);
    this.query = {
      name: this.query.name,
      text,
      params: [],
    };
  }

  enterParamName(ctx: ParamNameContext) {
    this.currentParam = {
      name: ctx.text,
      selection: undefined,
    };
  }

  enterScalarParamName(ctx: ScalarParamNameContext) {
    const required = !!ctx.REQUIRED_MARK();
    const name = ctx.ID().text;

    this.currentParam = {
      name,
      required,
    };
  }

  exitParam(ctx: ParamContext) {
    const defLoc = {
      a: ctx.start.startIndex,
      b: ctx.stop!.stopIndex,
      line: ctx.start.line,
      col: ctx.start.charPositionInLine,
    };
    this.currentParam.location = defLoc;
    this.currentParam.selection = this.currentSelection as ParamSelection;
    this.query.params!.push(this.currentParam as Param);
    this.currentSelection = {};
    this.currentParam = {};
  }

  enterScalarParam() {
    this.currentSelection = {
      type: ParamType.Scalar,
    };
  }

  enterPickParam() {
    this.currentSelection = {
      type: ParamType.Object,
      keys: [],
    };
  }

  enterArrayPickParam() {
    this.currentSelection = {
      type: ParamType.ObjectArray,
      keys: [],
    };
  }

  enterArrayParam() {
    this.currentSelection = {
      type: ParamType.ScalarArray,
    };
  }

  enterPickKey(ctx: PickKeyContext) {
    assert('keys' in this.currentSelection);

    const required = !!ctx.REQUIRED_MARK();
    const name = ctx.ID().text;

    this.currentSelection.keys!.push({ name, required });
  }
}

function parseText(
  text: string,
  queryName: string = 'query',
): { query: Query; events: ParseEvent[] } {
  const logger = new Logger();
  const inputStream = CharStreams.fromString(text);
  const lexer = new QueryLexer(inputStream);
  lexer.removeErrorListeners();
  lexer.addErrorListener(logger);
  const tokenStream = new CommonTokenStream(lexer);
  const parser = new QueryParser(tokenStream);
  parser.removeErrorListeners();
  parser.addErrorListener(logger);

  const tree = parser.input();

  const listener = new ParseListener(queryName, logger);
  ParseTreeWalker.DEFAULT.walk(listener as QueryParserListener, tree);

  return {
    query: listener.query as any,
    events: logger.parseEvents,
  };
}

export default parseText;
