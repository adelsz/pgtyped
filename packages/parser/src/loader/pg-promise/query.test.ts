import { parseTextPgPromise as parse } from './query';

test('scalar named param', () => {
  const query = `select * from users where id = $(id) and title = $(title)`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});

test('scalar index param', () => {
  const query = `select * from users where id = $1 and title = $2`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});

test('scalar named param nullable', () => {
  const query = `select * from users where id = $(id)/*nullable*/ and title = $(title)`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});

test('scalar index param nullable', () => {
  const query = `select * from users where id = $1 and title = $2/*nullable*/`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});

test('array param', () => {
  const query = `select * from users where id in ($(ids:list))`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});

test('array param nullable', () => {
  const query = `select * from users where id in ($(ids:list)/*nullable*/)`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});

test('identifier param', () => {
  const query = `select * from $(table:name)`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});

test('identifier param short', () => {
  const query = `select * from $(table~)`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});

test('raw param', () => {
  const query = `select * from $(table:raw)`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});

test('raw param short', () => {
  const query = `select * from $(table^)`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});

test('ignore line comments', () => {
  const query = `
select * from users where id in ($(ids:list))
-- this should be ignored: $(foo)
-- also this: 'hi hey"
-- and this: åäöÅÄÖ
`;

  const result = parse(query);
  expect(result).toMatchSnapshot();
});
