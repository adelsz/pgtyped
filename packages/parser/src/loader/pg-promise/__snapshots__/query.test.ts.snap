// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array param 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 33,
          "b": 43,
          "col": 33,
          "line": 1,
        },
        "name": "ids",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "scalar_array",
        },
      },
    ],
    "text": "select * from users where id in ($(ids:list))",
  },
}
`;

exports[`array param nullable 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 33,
          "b": 55,
          "col": 33,
          "line": 1,
        },
        "name": "ids",
        "nullable": true,
        "required": true,
        "selection": Object {
          "type": "scalar_array",
        },
      },
    ],
    "text": "select * from users where id in ($(ids:list)/*nullable*/)",
  },
}
`;

exports[`identifier param 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 14,
          "b": 26,
          "col": 14,
          "line": 1,
        },
        "name": "table",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "identifier",
        },
      },
    ],
    "text": "select * from $(table:name)",
  },
}
`;

exports[`identifier param short 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 14,
          "b": 22,
          "col": 14,
          "line": 1,
        },
        "name": "table",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "identifier",
        },
      },
    ],
    "text": "select * from $(table~)",
  },
}
`;

exports[`ignore line comments 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 34,
          "b": 44,
          "col": 33,
          "line": 2,
        },
        "name": "ids",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "scalar_array",
        },
      },
    ],
    "text": "
select * from users where id in ($(ids:list))
-- this should be ignored: $(foo)
-- also this: 'hi hey\\"
-- and this: åäöÅÄÖ",
  },
}
`;

exports[`raw param 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 14,
          "b": 25,
          "col": 14,
          "line": 1,
        },
        "name": "table",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "raw",
        },
      },
    ],
    "text": "select * from $(table:raw)",
  },
}
`;

exports[`raw param short 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 14,
          "b": 22,
          "col": 14,
          "line": 1,
        },
        "name": "table",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "raw",
        },
      },
    ],
    "text": "select * from $(table^)",
  },
}
`;

exports[`scalar index param 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 31,
          "b": 32,
          "col": 31,
          "line": 1,
        },
        "name": "1",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "scalar",
        },
      },
      Object {
        "location": Object {
          "a": 46,
          "b": 47,
          "col": 46,
          "line": 1,
        },
        "name": "2",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "scalar",
        },
      },
    ],
    "text": "select * from users where id = $1 and title = $2",
  },
}
`;

exports[`scalar index param nullable 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 31,
          "b": 32,
          "col": 31,
          "line": 1,
        },
        "name": "1",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "scalar",
        },
      },
      Object {
        "location": Object {
          "a": 46,
          "b": 59,
          "col": 46,
          "line": 1,
        },
        "name": "2",
        "nullable": true,
        "required": true,
        "selection": Object {
          "type": "scalar",
        },
      },
    ],
    "text": "select * from users where id = $1 and title = $2/*nullable*/",
  },
}
`;

exports[`scalar named param 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 31,
          "b": 35,
          "col": 31,
          "line": 1,
        },
        "name": "id",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "scalar",
        },
      },
      Object {
        "location": Object {
          "a": 49,
          "b": 56,
          "col": 49,
          "line": 1,
        },
        "name": "title",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "scalar",
        },
      },
    ],
    "text": "select * from users where id = $(id) and title = $(title)",
  },
}
`;

exports[`scalar named param nullable 1`] = `
Object {
  "events": Array [],
  "query": Object {
    "name": "query",
    "params": Array [
      Object {
        "location": Object {
          "a": 31,
          "b": 47,
          "col": 31,
          "line": 1,
        },
        "name": "id",
        "nullable": true,
        "required": true,
        "selection": Object {
          "type": "scalar",
        },
      },
      Object {
        "location": Object {
          "a": 61,
          "b": 68,
          "col": 61,
          "line": 1,
        },
        "name": "title",
        "nullable": false,
        "required": true,
        "selection": Object {
          "type": "scalar",
        },
      },
    ],
    "text": "select * from users where id = $(id)/*nullable*/ and title = $(title)",
  },
}
`;
